pkg/apis/discovery/validation/validation.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
+	metavalidation "k8s.io/apimachinery/pkg/apis/meta/v1/validation"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	apivalidation "k8s.io/kubernetes/pkg/apis/core/validation"
+	"k8s.io/kubernetes/pkg/apis/discovery"
+)
+
+var (
+	supportedAddressTypes  = sets.NewString(string(discovery.AddressTypeIP))
+	supportedPortProtocols = sets.NewString(string(api.ProtocolTCP), string(api.ProtocolUDP), string(api.ProtocolSCTP))
+	maxTopologyLabels      = 16
+	maxAddresses           = 100
+	maxPorts               = 100
+	maxEndpoints           = 1000
+)
+
+func ValidateEndpointSlice(endpointSlice *discovery.EndpointSlice) field.ErrorList {
+	validateEndpointSliceName := apimachineryvalidation.NameIsDNSSubdomain
+	allErrs := apivalidation.ValidateObjectMeta(&endpointSlice.ObjectMeta, true, validateEndpointSliceName, field.NewPath("metadata"))
+
+	addrType := discovery.AddressType("")
+	if endpointSlice.AddressType == nil {
+		allErrs = append(allErrs, field.Required(field.NewPath("addressType"), ""))
+	} else if !supportedAddressTypes.Has(string(*endpointSlice.AddressType)) {
+		allErrs = append(allErrs, field.NotSupported(field.NewPath("addressType"), *endpointSlice.AddressType, supportedAddressTypes.List()))
+	} else {
+		addrType = *endpointSlice.AddressType
+	}
+
+	allErrs = append(allErrs, validateEndpoints(endpointSlice.Endpoints, addrType, field.NewPath("endpoints"))...)
+	allErrs = append(allErrs, validatePorts(endpointSlice.Ports, field.NewPath("ports"))...)
+
+	return allErrs
+}
+
+func ValidateEndpointSliceUpdate(newEndpointSlice, oldEndpointSlice *discovery.EndpointSlice) field.ErrorList {
+	allErrs := ValidateEndpointSlice(newEndpointSlice)
+
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(*newEndpointSlice.AddressType, *oldEndpointSlice.AddressType, field.NewPath("addressType"))...)
+
+	return allErrs
+}
+
+func validateEndpoints(endpoints []discovery.Endpoint, addrType discovery.AddressType, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(endpoints) > maxEndpoints {
+		allErrs = append(allErrs, field.TooMany(fldPath, endpoints, maxEndpoints))
```

Comment:
Maybe we should terminate here, rather than process a list we know is too long - what think?  If so, same in other places.